vllm/attention/backends/ipex_attn.py

```diff
@@ -0,0 +1,393 @@
+            attn_metadata: Metadata for attention.
+        Returns:
+            shape = [num_tokens, num_heads * head_size]
+        """
+        assert kv_scale == 1.0
+        num_tokens, hidden_size = query.shape
+        # Reshape the query, key, and value tensors.
+        query = query.view(-1, self.num_heads, self.head_size)
+        key = key.view(-1, self.num_kv_heads, self.head_size)
+        value = value.view(-1, self.num_kv_heads, self.head_size)
+
+        if kv_cache is not None:
+            key_cache, value_cache = self.split_kv_cache(
+                kv_cache, self.num_kv_heads, self.head_size)
+            ipex_ops.reshape_and_cache(
+                key,
+                value,
+                key_cache,
+                value_cache,
+                attn_metadata.slot_mapping.flatten(),
+                self.kv_cache_dtype,
+                kv_scale,
+            )
+
+        if attn_metadata.is_prompt:
+            assert attn_metadata.seq_lens is not None
+            if (kv_cache is None or attn_metadata.block_tables.numel() == 0):
+                if self.num_kv_heads != self.num_heads:
+                    key = key.repeat_interleave(self.num_queries_per_kv, dim=1)
+                    value = value.repeat_interleave(self.num_queries_per_kv,
+                                                    dim=1)
+
+                if attn_metadata.attn_bias is None:
+                    if self.alibi_slopes is not None:
+                        att_masks = _make_alibi_bias(
+                            self.alibi_slopes, query.dtype,
+                            attn_metadata.seq_lens)  # type: ignore
+                    elif self.sliding_window is not None:
+                        att_masks = _make_sliding_window_bias(
+                            attn_metadata.seq_lens, self.sliding_window,
+                            query.dtype)  # type: ignore
+                    else:
+                        if self.fuse_batch:
+                            att_masks = _make_sliding_window_bias(
+                                attn_metadata.seq_lens,
+                                None,
+                                dtype=query.dtype)
+                        else:
+                            att_masks = [None] * len(attn_metadata.seq_lens)
+                    attn_metadata.attn_bias = att_masks
+
+                if self.fuse_batch:
+                    out = torch.empty(
+                        (num_tokens, self.num_heads, self.head_size),
+                        dtype=query.dtype,
+                        device=query.device)
+                    tmp = [0]
+                    tmp.extend(attn_metadata.seq_lens)
+                    seqlen = torch.tensor(tmp)
+                    max_seqlen = max(attn_metadata.seq_lens)
+                    seqlen_q = torch.cumsum(seqlen,
+                                            dim=0).to(device=query.device)
```

Comment:
While this is not necessary, could we move this to the model runner so that it's not executed every attention layer?